gap/base/projective.gi

```diff
@@ -16,50 +16,53 @@
 ##
 #############################################################################
 
+# Assumes <m> is a non-empty square matrix defined over a field. <m> defines a
+# projective one iff it is a non-zero scalar multiple of the identity matrix.
 InstallGlobalFunction( IsOneProjective,
-  function(el)
-    local s, n, i, j, zero;
-    n := NrRows(el);
-    Assert(0, n = NrCols(el));
-    s := el[1,1];
-    if IsZero(s) then return false; fi;
-    zero := Zero(s);
-    for i in [1..n] do
-        if el[i,i] <> s then return false; fi;
-        for j in [1..n] do
-            if i <> j and el[i,j] <> zero then return false; fi;
-        od;
-    od;
-    return true;
+  function(m)
+    Assert(1, NrRows(m) > 0 and NrRows(m) = NrCols(m));
+    return not IsZero(m[1,1]) and RECOG.IsScalarMat(m) <> false;
   end );
 
+# Assumes <a> and <b> are non-empty equal-dimension square matrices over equal
+# fields, and that the first row of <a> contains a non-zero entry.
+#
+# <a> and <b> define equal projective elements, or are "equal modulo scalars",
+# iff there exists a (non-zero) scalar <s> in the field such that <s * a = b>.
 InstallGlobalFunction( IsEqualProjective,
-  function(a,b)
-    local p,s,i;
-    p := PositionNonZero(a[1]);
-    s := b[1,p];
-    if IsZero(s) then return false; fi;
-    s := s / a[1,p];
-    for i in [1..Length(a)] do
-        if s*a[i] <> b[i] then return false; fi;
-    od;
-    return true;
+  function(a, b)
+    local n, p, s;
+    n := NrRows(a);
+    Assert(1, n > 0 and n = NrRows(b) and n = NrCols(a) and n = NrCols(b));
+    p := First([1..n], i -> not IsZero(a[1,i])); # Find non-zero entry in <a[1]>
+    Assert(1, p <> fail and BaseDomain(a) = BaseDomain(b));
+    s := b[1,p] / a[1,p]; # The unique scalar <s> with <s * a[1,p] = b[1,p]>.
+    return not IsZero(s)
+           and ForAll([1..n], i -> ForAll([1..n], j -> s * a[i,j] = b[i,j]));
```

Comment:
Note that using ForAll like this is a bit slower than using an explicit unrolled loop; this is mostly due to the overhead for calling a GAP function for every single matrix entry (in the worst case).
And, sadly, switching from row-based to entry based checks can *totally* kill performance.

 We can verify all this with a micro benchmark:
```
gap> n:=100;; a:=IdentityMat(n, GF(3));;
gap> s:=Z(3)^0;;
gap>
gap> a:=a*s;; # uncompressed FFE matrix
gap> for i in [1..1000] do for i in [1..n] do for j in [1..n] do x := a[i,j] = s*a[i,j]; od; od; od; time;
1696
gap> for i in [1..1000] do ForAll([1..n], i -> ForAll([1..n], j -> a[i,j] = s*a[i,j])); od; time;
1940
gap> for i in [1..1000] do for i in [1..n] do x := a[i] = s*a[i]; od; od; time;
36
gap> for i in [1..1000] do ForAll([1..n], i -> a[i] = s*a[i]); od; time;
36
gap> for i in [1..1000] do x := a = s*a; od; time;
20
gap>
gap> ConvertToMatrixRep(a);; # compressed matrix
gap> for i in [1..1000] do for i in [1..n] do for j in [1..n] do x := a[i,j] = s*a[i,j]; od; od; od; time;
1692
gap> for i in [1..1000] do ForAll([1..n], i -> ForAll([1..n], j -> a[i,j] = s*a[i,j])); od; time;
2048
gap> for i in [1..1000] do for i in [1..n] do x := a[i] = s*a[i]; od; od; time;
32
gap> for i in [1..1000] do ForAll([1..n], i -> a[i] = s*a[i]); od; time;
48
gap> for i in [1..1000] do x := a = s*a; od; time;
80
```
I therefore suggest that we check for `IsRowListMatrix`, and try to fallback to the row based approach if possible; and if that is not possible, then we use `s*a = b` instead (which needs to create a complete new matrix, which causes the extra overhead seen in the last test case above).

For optimal performance, one really needs to taylor the code to the matrix representation. That would suggest turning this (and the other function in this file) into operations, allowing the installation of additional methods for other IsMatrixObj implementations. *Except* that this will slow down things for "classic" `IsMatrix` plist-of-plist matrices, which trigger method selection over head... But arguably, we shouldn't be seeing these here anyway, only compressed FFE matrices... So perhaps this operation approach would be the cleanest way to achieve what I suggest above (and we could even have a method which triggers for plist inputs, and prints out a performance warning)

Reply:
I've pushed a second commit taking on board some of what you've said. If it's what you meant, then I can squash it into the main commit, if not I can make some changes.

I agree that using operations should be the best idea for optimal performance, but I don't want to do it with this PR. I've done as much work on this file as I want to at the moment, and since (at least I feel) this  PR improves things, I think it would be nice to get it merged before I/you/someone else makes future additional improvements 🙂